### PR TITLE
RATEST-223: Add Edit Vitals Page for Vitals and Triaging workflow test

### PR DESCRIPTION
**Pull request by Irenyak1**
Ticket  [RATEST-223](https://issues.openmrs.org/browse/RATEST-223)

Description ->-> Added  Edit Vitals Page for Vitals and Triaging workflow test

### CODE REVIEW

File: ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java [hunk 1]
```diff
@@ -86,6 +86,10 @@
     private static final By DELETE_PATIENT_CANCEL  = By.cssSelector("#delete-patient-creation-dialog > div.dialog-content > button.cancel");
     private static final By EDIT_PATIENT_LINK = By.cssSelector("a[id='application.registrationapp.summary.editPatientLink']");
     private static final String REASON = "patient nolonger needed";
+    private static final By EDIT_VITALS_ICON = By.cssSelector("#encountersList i.editEncounter.edit-action.icon-pencil");
+    private static final By VIEW_VITALS_ICON  = By.cssSelector("#encountersList i.viewEncounter.view-action.icon-file-alt");
+    private static final By DELETE_VITALS_ICON = By.cssSelector("#encountersList i.deleteEncounterId.delete-action.icon-remove");
+    private static final By COMFIRM_DELETE_BUTTON  = By.cssSelector("#delete-encounter-dialog button.confirm.right");
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:92]: Are these elements on ClinicianFacingPatientDashboardPage or PatientVisitsDashboardPage?

```js
private static final By EDIT_VITALS_ICON = By.cssSelector("#encountersList i.editEncounter.edit-action.icon-pencil");
    private static final By VIEW_VITALS_ICON  = By.cssSelector("#encountersList i.viewEncounter.view-action.icon-file-alt");
    private static final By DELETE_VITALS_ICON = By.cssSelector("#encountersList i.deleteEncounterId.delete-action.icon-remove");
```

Comment (Irenyak1) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:92]: > @Irenyak1 have mercy with the grand parent and take back the grand sons to their father, imagine all the grand sons of ClinicianFacingPatientDashboardPage are hosted at his little house, don't you think it will be overwhelming to him?

@kdaud  I definitely did not want to have these grand sons at their grand father's house but seems the father does not want to take the responsibility! :) . Man I tried to implement such workarounds before just to make sure the  PatientVisitDashboardPage serves its purspose but all was in vain. However, let me try again with your suggestion and see it will succeed.. Thanks.

Comment (Irenyak1) [reply ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java:92]: But Actually I am waiting for your feedback about the sync.

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVitalsPage.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
+package org.openmrs.reference.page;
+
+import org.openmrs.uitestframework.page.Page;
```
Comment (sherrif10) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVitalsPage.java:3]: Change this to import from org.openmrs.contrib.qaframework.helper.page
```suggestion
import org.openmrs.contrib.qaframework.helper.page.```

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java [hunk 3]
```diff
@@ -44,9 +49,29 @@ public PatientVisitsDashboardPage(Page parent, WebElement waitForStaleness) {
     public String getPageUrl() {
         return "coreapps/patientdashboard/patientDashboard.page";
     }
+    
+    public PatientCaptureVitalsPage goToPatientCaptureVitalsPage() {
+        waiter.until(ExpectedConditions.presenceOfElementLocated(CAPTURE_VITALS));
+        clickOnLast(CAPTURE_VITALS);
```
Comment (sherrif10) [inline ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:55]: @kdaud  she have added it here, @Irenyak1  is this how the workflow is supposed to flow ?

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:55]: Thanks @sherrif10 for pointing this out. It actually makes the return type to be true for the method call.

One thing I would recommend here is to rather use clickOn() when clicking on the page elements.

Comment (Irenyak1) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:55]: > Thanks @sherrif10 for pointing this out. It actually makes the return type to be true for the method call.
> 
> @Irenyak1 one thing I would recommend here is to rather use `clickOn()` when clicking on the page element.

@kdaud  I get it. Are there any other comments I should address before I make another push?

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java]: > Are there any other comments I should address before I make another push?

If I was @Irenyak1, I would not always wait reviewers to put all their comments on the PR then I address them at once. Why? 
Reviewing a PR is a process and can take more time when the PR author is not addressing the pending reviews waiting for all the reviews at once. 

I learnt addressing any review on my PR immediately and push the requested changes. Then wait for any other review(s) and if any I immediately address them as well. This saves my time when working on a PR especially when reviewers are giving quick feedback. 

I learnt this logic from @dkayiwa and am now a consultant to my self.

Comment (Irenyak1) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java]: @kdaud of course, I can not sit and wait. I attended to all the reviews long time ago. Just waiting for others if any, as I do something else. No time to Waste!

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java [hunk 4]
```diff
@@ -44,9 +49,29 @@ public PatientVisitsDashboardPage(Page parent, WebElement waitForStaleness) {
...
+    
+    public EditVitalsPage goToEditVitalsPage() {
+        waiter.until(ExpectedConditions.presenceOfElementLocated(EDIT_VITALS_ICON));
+        clickOn(EDIT_VITALS_ICON);
+        return new EditVitalsPage(this);
+    }
 
-    public void goToCaptureVitals() {
-        findElement(CAPTURE_VITALS).click();
+    public void goToViewVitalsPage() {
```
[the remark this answers is not in the document]
Comment (Irenyak1) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java]: @kdaud done.

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java [hunk 5]
```diff
@@ -44,9 +49,29 @@ public PatientVisitsDashboardPage(Page parent, WebElement waitForStaleness) {
...
-        findElement(CAPTURE_VITALS).click();
+    public void clickOnViewVitalsIcon() {
+        waiter.until(ExpectedConditions.presenceOfElementLocated(VIEW_VITALS_ICON));
+        clickOn(VIEW_VITALS_ICON);
+    }
+
+    public void deleteVitals() {
+        waiter.until(ExpectedConditions.presenceOfElementLocated(DELETE_VITALS_ICON));
+        clickOn(DELETE_VITALS_ICON);
+        waitForElement(COMFIRM_DELETE_BUTTON);
```
Comment (sherrif10) [inline ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:73]: What happens when you dont have this line 73

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:73]: So ideally which error does it return when it takes long to be clicked

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java:73]: When a delete action is invoked on the delete button, its a good design to have it throw a confirm dialogue for the user to confirm the action. There is a milliseconds taken for the confirm dialogue to be rendered on the UI and so requires to invoke waitForElement() function for the next action to be effective.
